File: 01-collatz/collatz.go
package main

import (
    "fmt"
    "os"
)

func collatz(n int) int {
    cnt := 0
    for n > 1 {
        if n % 2 > 0 {
            n = 3 * n + 1
        } else {
            n = n / 2
        }
        cnt++
    }
    return cnt
}

func main() {
    var maxn int
    fmt.Sscanf(os.Getenv("MAXN"), "%d", &maxn)
    sum := 0
    for i := 1; i <= maxn; i++ {
        sum += collatz(i)
    }
    fmt.Printf("sum=%d, avg=%f\n", sum, float64(sum) / float64(maxn))
}

File: 01-collatz/collatz.c
#include <stdio.h>
#include <stdlib.h>

int collatz(long long x) {
    int cnt = 0;
    while (x > 1) {
        x = x % 2 ? 3 * x + 1 : x / 2;
        cnt++;
    }
    return cnt;
}

int main(void) {
    int n, i, sum = 0;
    sscanf(getenv("MAXN"), "%d", &n);
    for (i = 1; i <= n; i++) {
        sum += collatz(i);
    }
    printf("sum=%d, avg=%.2f\n", sum, sum / (double) n);
}
